packages/core/core-flows/src/order/workflows/create-return.ts

```diff
@@ -0,0 +1,305 @@
+  const inputItemIds = inputItems.map((i) => i.id)
+  const diff = arrayDifference(inputItemIds, orderItemIds)
+
+  if (diff.length) {
+    throw new MedusaError(
+      MedusaError.Types.INVALID_DATA,
+      `Items with ids ${diff.join(", ")} does not exist in order with id ${
+        order.id
+      }.`
+    )
+  }
+}
+
+function validateReturnReasons(
+  {
+    order_id,
+    inputItems,
+  }: {
+    order_id: string
+    inputItems: OrderWorkflow.CreateOrderReturnWorkflowInput["items"]
+  },
+  { container }
+) {
+  const reasonIds = inputItems.map((i) => i.reason_id).filter(Boolean)
+
+  if (!reasonIds.length) {
+    return
+  }
+
+  const remoteQuery = container.resolve(ContainerRegistrationKeys.REMOTE_QUERY)
+
+  const returnReasons = remoteQuery({
+    entry_point: "return_reasons",
+    fields: ["return_reason_children.*"],
+    variables: { id: [inputItems.map((item) => item.reason_id)] },
+  })
+
+  const reasons = returnReasons.map((r) => r.id)
+  const hasUnusableReasons = reasons.filter(
+    (reason) => reason.return_reason_children.length === 0
+  )
+  const hasUnExistingReasons = arrayDifference(reasonIds, reasons)
+
+  if (hasUnExistingReasons.length) {
+    throw new MedusaError(
+      MedusaError.Types.INVALID_DATA,
+      `Return reason with id ${hasUnExistingReasons.join(
+        ", "
+      )} does not exists.`
+    )
+  }
+
+  if (hasUnusableReasons.length()) {
+    throw new MedusaError(
+      MedusaError.Types.INVALID_DATA,
+      `Cannot apply return reason category with id ${hasUnusableReasons.join(
+        ", "
+      )} to order with id ${order_id}.`
+    )
+  }
+}
+
+function prepareShippingMethodData({
+  orderId,
+  inputShippingOption,
+  returnShippingOption,
+}: {
+  orderId: string
+  inputShippingOption: OrderWorkflow.CreateOrderReturnWorkflowInput["return_shipping"]
+  returnShippingOption: ShippingOptionDTO & {
+    calculated_price: { calculated_amount: number }
+  }
+}) {
+  const obj: CreateOrderShippingMethodDTO = {
+    name: returnShippingOption.name,
+    order_id: orderId,
+    shipping_option_id: returnShippingOption.id,
+    amount: 0,
+    data: {},
+    // Computed later in the flow
+    tax_lines: [],
+    adjustments: [],
+  }
+
+  if (isDefined(inputShippingOption.price) && inputShippingOption.price >= 0) {
+    obj.amount = inputShippingOption.price
+  } else {
+    if (returnShippingOption.price_type === "calculated") {
+      // TODO: retrieve calculated price and assign to amount
+    } else {
+      obj.amount = returnShippingOption.calculated_price.calculated_amount
+    }
+  }
+
+  return obj
+}
+
+function validateCustomRefundAmount({
+  order,
+  refundAmount,
+}: {
+  order: Pick<OrderDTO, "item_total">
+  refundAmount?: number
+}) {
+  // validate that the refund prop input is less than order.item_total (item total)
+  // TODO: Probably this amount should be retrieved from the payments linked to the order
+  if (refundAmount && MathBN.gt(refundAmount, order.item_total)) {
```

Comment:
qusetion: We don't have a `refundable_amount` on the order? If not, I think we should consider it

Reply:
yes for now we don't cc @carlos-r-l-rodrigues I think we discuss it as the refundable amount is basically everything that has been paid minus refunded